File: src/models/audio_digest.py
"""AudioDigest model for tracking audio digest generation.

This module provides models for tracking text-to-speech generation
of digest content, supporting multiple TTS providers (OpenAI, ElevenLabs, etc.)
with configurable voice and speed options.
"""

from datetime import UTC, datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy import (
    Column,
    DateTime,
    Enum as SQLEnum,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
)
from sqlalchemy.orm import relationship

from src.models.base import Base


class AudioDigestStatus(str, Enum):
    """Status of audio digest generation."""

    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class AudioDigest(Base):
    """Audio digest database model.

    Tracks text-to-speech generation of digest content.
    Each digest can have multiple audio versions with different
    voice/speed configurations.
    """

    __tablename__ = "audio_digests"

    id = Column(Integer, primary_key=True, autoincrement=True)
    digest_id = Column(Integer, ForeignKey("digests.id"), nullable=False, index=True)

    # Generation config
    voice = Column(String(50), nullable=False)  # e.g., "nova", "alloy"
    speed = Column(Float, default=1.0)
    provider = Column(String(50), default="openai")  # TTS provider

    # Status tracking
    status = Column(
        SQLEnum(AudioDigestStatus, values_callable=lambda x: [e.value for e in x]),
        default=AudioDigestStatus.PENDING,
        nullable=False,
        index=True,
    )
    error_message = Column(Text, nullable=True)

    # Output
    audio_url = Column(String(500), nullable=True)  # Storage path
    duration_seconds = Column(Float, nullable=True)
    file_size_bytes = Column(Integer, nullable=True)

    # Analytics
    text_char_count = Column(Integer, nullable=True)
    chunk_count = Column(Integer, nullable=True)

    # Timestamps
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(UTC))
    completed_at = Column(DateTime, nullable=True)

    # Relationships
    digest = relationship("Digest", back_populates="audio_digests")


# --- Pydantic Schemas ---


class AudioDigestCreate(BaseModel):
    """Schema for creating a new audio digest."""

    voice: str = Field(default="nova", description="Voice to use for TTS")
    speed: float = Field(default=1.0, ge=0.25, le=4.0, description="Playback speed")
    provider: str = Field(default="openai", description="TTS provider")


class AudioDigestResponse(BaseModel):
    """Schema for audio digest API responses."""

    id: int
    digest_id: int
    voice: str
    speed: float
    provider: str
    status: AudioDigestStatus
    error_message: str | None = None
    audio_url: str | None = None
    duration_seconds: float | None = None
    file_size_bytes: int | None = None
    text_char_count: int | None = None
    chunk_count: int | None = None
    created_at: datetime
    completed_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class AudioDigestListItem(BaseModel):
    """Lightweight schema for listing audio digests."""

    id: int
    digest_id: int
    voice: str
    provider: str
    status: AudioDigestStatus
    duration_seconds: float | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: web/src/types/audio-digest.ts
/**
 * Audio Digest Types
 *
 * TypeScript interfaces for audio digest entities.
 * Audio digests are TTS-generated audio from Digests (single-voice narration).
 *
 * Unlike podcasts which require script generation and review, audio digests
 * provide direct digest-to-audio conversion with configurable voice and speed.
 *
 * @see Backend model: src/models/audio_digest.py
 * @see Backend API: src/api/audio_digest_routes.py
 */

/**
 * Audio digest generation status
 *
 * Tracks the audio digest through the generation pipeline.
 */
export type AudioDigestStatus =
  | "pending" // Created, queued for processing
  | "processing" // Currently generating audio
  | "completed" // Successfully generated
  | "failed" // Generation failed

/**
 * TTS voice provider options
 *
 * Available text-to-speech providers for audio generation.
 */
export type AudioDigestProvider = "openai" | "elevenlabs"

/**
 * Available voice options
 *
 * Voice presets available for audio digest narration.
 * These map to provider-specific voice IDs on the backend.
 */
export type AudioDigestVoice =
  | "nova" // Warm female (OpenAI default)
  | "onyx" // Deep male
  | "echo" // Natural male
  | "shimmer" // Expressive female
  | "alloy" // Neutral
  | "fable" // Storytelling

/**
 * Audio digest list item (lightweight view)
 *
 * Used in list views and tables.
 * Field names use snake_case to match the Python backend API.
 */
export interface AudioDigestListItem {
  /** Unique identifier */
  id: number
  /** Source digest ID */
  digest_id: number
  /** Voice used for synthesis */
  voice: string
  /** Speech speed multiplier */
  speed: number
  /** TTS provider used */
  provider: string
  /** Current status */
  status: AudioDigestStatus
  /** Audio duration in seconds (null if not completed) */
  duration_seconds: number | null
  /** File size in bytes (null if not completed) */
  file_size_bytes: number | null
  /** When created */
  created_at: string
  /** When completed (null if not completed) */
  completed_at: string | null
  /** Error message if failed */
  error_message: string | null
}

/**
 * Audio digest full details
 *
 * Complete information including audio URL and analytics.
 * Field names use snake_case to match the Python backend API.
 */
export interface AudioDigestDetail extends AudioDigestListItem {
  /** URL to audio file (null if not completed) */
  audio_url: string | null
  /** Audio format (always "mp3") */
  audio_format: string
  /** Number of characters in prepared text */
  text_char_count: number | null
  /** Number of TTS synthesis chunks */
  chunk_count: number | null
}

/**
 * Audio digest statistics
 *
 * Aggregate statistics for the audio digests page.
 */
export interface AudioDigestStatistics {
  /** Total number of audio digests */
  total: number
  /** Currently generating */
  generating: number
  /** Successfully completed */
  completed: number
  /** Failed generation */
  failed: number
  /** Total audio duration in seconds */
  total_duration_seconds: number
  /** Count by voice */
  by_voice: Record<string, number>
  /** Count by provider */
  by_provider: Record<string, number>
}

/**
 * Request to create an audio digest
 *
 * Configuration for audio digest generation.
 */
export interface CreateAudioDigestRequest {
  /** Voice preset to use */
  voice?: AudioDigestVoice
  /** Speech speed multiplier (0.5 to 2.0) */
  speed?: number
  /** TTS provider */
  provider?: AudioDigestProvider
}

/**
 * Response from create audio digest endpoint
 */
export interface CreateAudioDigestResponse {
  /** Status message */
  status: string
  /** Human-readable message */
  message: string
  /** Created audio digest ID */
  audio_digest_id: number
}

/**
 * Filters for audio digest list
 */
export interface AudioDigestFilters {
  /** Filter by status */
  status?: AudioDigestStatus
  /** Filter by provider */
  provider?: string
  /** Filter by voice */
  voice?: string
  /** Limit number of results */
  limit?: number
  /** Offset for pagination */
  offset?: number
  /** Sort field */
  sort_by?: string
  /** Sort order */
  sort_order?: "asc" | "desc"
}

/**
 * Digest available for audio generation
 *
 * Lightweight digest info for the source selector dropdown.
 */
export interface AvailableDigest {
  /** Digest ID */
  id: number
  /** Digest title */
  title: string | null
  /** Digest type (daily, weekly) */
  digest_type: string
  /** Period start date */
  period_start: string
  /** Period end date */
  period_end: string
  /** Current status */
  status: string
  /** When created */
  created_at: string
}

/**
 * Voice option for the selector dropdown
 */
export interface VoiceOption {
  /** Voice ID value */
  value: AudioDigestVoice
  /** Display label */
  label: string
  /** Voice description */
  description: string
}

/**
 * Provider option for the selector dropdown
 */
export interface ProviderOption {
  /** Provider ID value */
  value: AudioDigestProvider
  /** Display label */
  label: string
}
